### PR TITLE
catabalancer/systats: add bw usage info

### PR DESCRIPTION
**Pull request by emranemran**
TODO: need a separate PR to add `vnstat` to our pods
~~TODO: read NIC speed directly from /proc filesystem (need to confirm if this is accurate)~~

### CODE REVIEW

File: balancer/catabalancer/sysstats.go [hunk 1]
```diff
@@ -63,3 +95,31 @@ func GetSystemUsage() (SystemUsage, error) {
...
+
+	var data BandwidthData
+	err = json.Unmarshal([]byte(output), &data)
+	if err != nil {
+		return -1, err
+	}
+
+	totalBW := float64(data.Rx.Bytes) + float64(data.Tx.Bytes)
+	// TODO: is our NICs 20Gbps?? 
+	totalBWPercent := (totalBW / (20 * 1e9)) * 100
```
Comment (emranemran) [inline balancer/catabalancer/sysstats.go:122]: hmmmm maybe not? 
```
root@prod-catalyst-0:/# cat /sys/class/net/eth0/speed
10000
```

Comment (thomshutt) [reply balancer/catabalancer/sysstats.go:122]: Yeah, 10 sounds right but it varies by machine

Comment (thomshutt) [reply balancer/catabalancer/sysstats.go:122]: You can confirm via the Datapacket website though

File: balancer/catabalancer/sysstats.go [hunk 2]
```diff
@@ -63,3 +95,31 @@ func GetSystemUsage() (SystemUsage, error) {
 
 	return systemUsage, nil
 }
+
+func GetBandwidthUsage() (float64, error) {
+	config := VnstatConfig{
```
[the remark this answers is not in the document]
Comment (emranemran) [reply balancer/catabalancer/sysstats.go]: Yeah fixed this - forgot to remove the struct from a small refactor.

File: balancer/catabalancer/sysstats.go [hunk 3]
```diff
@@ -63,3 +95,31 @@ func GetSystemUsage() (SystemUsage, error) {
...
+		Interface: "eth0",
+		Args:      []string{"-tr", "2.5"},
+	}
+	args := append([]string{"--json", "--iface", config.Interface}, config.Args...)
+
+	cmd := exec.Command("vnstat", args...)
+	cmd.Env = os.Environ()
```
[the remark this answers is not in the document]
Comment (emranemran) [reply balancer/catabalancer/sysstats.go:111]: done

File: balancer/catabalancer/sysstats.go [hunk 4]
```diff
@@ -63,3 +95,31 @@ func GetSystemUsage() (SystemUsage, error) {
...
+		return -1, err
+	}
+
+	var data BandwidthData
+	err = json.Unmarshal([]byte(output), &data)
+	if err != nil {
+		return -1, err
+	}
+
+	totalBW := float64(data.Rx.Bytes) + float64(data.Tx.Bytes)
```
[the remark this answers is not in the document]
Comment (emranemran) [reply balancer/catabalancer/sysstats.go]: Yeah, fixed these calculations now.

File: balancer/catabalancer/sysstats.go [hunk 5]
```diff
@@ -63,3 +97,52 @@ func GetSystemUsage() (SystemUsage, error) {
...
+	if err != nil {
+		return -1, err
+	}
+	defer file.Close()
+	content := make([]byte, 64)
+	n, err := file.Read(content)
+	if err != nil {
+		return -1, err
+	}
+	speedStr := strings.TrimSpace(string(content[:n]))
```
Comment (mjh1) [inline balancer/catabalancer/sysstats.go:139]: it would be simpler and safer to just use `os.ReadFile()`

File: balancer/catabalancer/sysstats.go [hunk 6]
```diff
@@ -63,3 +97,52 @@ func GetSystemUsage() (SystemUsage, error) {
 
 	return systemUsage, nil
 }
+
+// Get bandwidth usage using the vnstat utility.
+// 'vnstat --json --iface en0 -tr 2.5' calculates traffic for given interface
```
Comment (mjh1) [inline balancer/catabalancer/sysstats.go:102]: does this call take 2.5s?

File: balancer/catabalancer/sysstats.go [hunk 7]
```diff
@@ -29,6 +34,28 @@ type SystemInfo struct {
 	DiskInfo []disk.UsageStat
 }
 
+type VnstatConfig struct {
```
Comment (mjh1) [inline balancer/catabalancer/sysstats.go:37]: Still need to remove this

**Review by mjh1 — CHANGES_REQUESTED** [review]

File: balancer/catabalancer/sysstats.go (hunk 5 above)
Comment (emranemran) [reply balancer/catabalancer/sysstats.go]: Updated.

File: balancer/catabalancer/sysstats.go (hunk 7 above)
Comment (emranemran) [reply balancer/catabalancer/sysstats.go]: oops, removed.

File: balancer/catabalancer/sysstats.go (hunk 6 above)
Comment (emranemran) [reply balancer/catabalancer/sysstats.go:97]: Yeah, the stats need to be averaged over a small duration. 1s is too inaccurate and 5s will block for too long. Perhaps an improvement for the future could be just getting the bandwidth estimates regularly in the background and the balancer logic can just use whatever was the last value recorded without blocking.

**Pull request merged**